API: Switch to NEP 50 behavior by default

seberg opened the pull request:
This is a WIP for discussion and visibility for switching to NEP 50 behavior by default.

As of writing, it fixes up mainly clear tests and one small hack to allow `isnan(2**100)` to pass.  That isn't perfectly clean maybe, but seems pragmatic.
(The individual commits make sense on their own)

There are some remaining tests that need fixing:
* Should `percentile(float32_arr, 30.)` return a float32 or float64 (currently it switches to float64), but we could add `np.result_type` to fix that.
  * Adding `np.result_type` isn't great, it would be nice to have something slightly cleaner.  I am not exactly sure what, though.  (A function to prep multiple arrays, something that @mdhaber would like anyway. A way to mark the array as representing a Python scalar?  But we want that to be used with caution, it would be weird to be used beyond temporaries)
* `can_cast(123, np.uint8)` isn't well defined right now.  I can live with that for the moment.
* `np.where()` doesn't do the weak-promotion, `np.concatenate()` does, so it should not be super hard to fix.
* `np.add(1., 2., dtype=np.int64)` should fail because of unsafe casting, but casting safety+scalars is still a bit of a mess.

I am not sure that all of these are easy, but will chip away.  OTOH some of these also look like things that might be OK as "known issues".

The main thing I would like is figure out `percentile`:  We won't get this right everywhere, but it would be good to know what our story is for such a function.

seberg committed:
MAINT,TST: Fix issues due to out-of-range integers

This contains two fixes:
1. a somewhat hackish solution to keep allowing very large integers
   for unary ufuncs.  This is partially wrong, but keeps isnan and
   other functions working (reducing fallout in the tests quite a bit)
2. Generally adapt changes that now fail for the same reason, this is
   mainly due to comparisons like `-1 < uint8` now failing in generally.
   That is fixable in principle, but it isn't super easy.

Neither is perfect, both are the pragmatic solutions at this time until
they proof to be too problematic for some reason.

(Getting rid of going to object for large integers is tougher and needs
deeper work.  It certainly is possible but too difficult at this stage:
we need to remove some old branches first.
We could, and maybe should, get rid of the `uint` step though, it is even
more of a trap I think!)

seberg committed:
MAINT: Robustify histogram unsigned-subtract and fix test for NEP 50

seberg committed:
TST: Adjust tests for direct NEP 50 changes

seberg committed:
TST: Straight forward NEP 50 fixup or longdouble+int scalar tests

seberg committed:
API: Switch to "weak" promotion state by default

seberg commented on numpy/lib/tests/test_histograms.py:
```diff
@@ -366,10 +364,9 @@ def do_precision_upper_bound(self, float_small, float_large):
 
-        # gh-10322 means that the type comes from arr - this may change
-        assert_equal(x_loc.dtype, float_small)
+        assert_equal(x_loc.dtype, float_large)
```
@eric-wieser these histogram changes look right to you?  What was previously rounded to the small bin, now ends up with the large dtype because it is explicit, I guess.

mhvk replied:
FWIW, I think this is logical as long as one doesn't special-case histogram in treating its arguments asymmetrically.

seberg committed:
STY: Remove unnecessary indentation level

This used to make a bit of sense in C99, but now just distracts

seberg committed:
API: Use weak scalar logic in where (and fix up error paths slightly)

seberg committed:
TST: Adept where test to NEP 50

seberg committed:
MAINT: Require `select` for NEP 50 support

This isn't quite ideally, it would be nice to grow specific code
for it (and actually, we currently allow some subclasses in other
paths, but I am planning to undo that).

seberg commented:
Really need some opinions on this.  I hacked `np.select()` to just check for int, float, complex explicitly for now, that is OK in practice.

`np.percentile` tests are more complicated, because either we have to make a clear decision that `np.percentile(float32_arr, 30.)` (I suspect also quantile).

This is an API decision that outlines the difficulty with NEP 50: where to draw the line and disregard the "weakly typed" float there.  For percentile it may yet be a bit simpler, but still:
1. This should maintain the weak scalar, this probably requires making sure that all math uses python scalars (or checking up-front and enforcing a Python float at the end).
   * I am really not sure how to design this correctly!
2. Draw the line here, quantile/percentile uses `q = np.asarray(q)` and unlike ufuncs *doesn't* use weak promotion (how to teach what does and what doesn't?)
3. For `percentile`/`quantile`, I guess it may make sense to ignore the interpolation precision.  We do currently return a `Fraction` when `q` is a Fraction though.  For some things, casting `q` is probably wrong.

Ping @mhvk, but really everyone should have an opinion on this, this is probably the most important thing to figure out for NEP 50 and I think a lot of people want NEP 50.

mhvk commented on numpy/lib/histograms.py:
```diff
@@ -348,13 +348,15 @@ def _unsigned_subtract(a, b):
-        return np.subtract(a, b, casting='unsafe', dtype=dt)
+        # signed to unsigned.  The input may be negative python integers so
+        # ensure we pass in arrays with the initial dtype (related to NEP 50).
+        return np.subtract(np.asarray(a, dtype=dt), np.asarray(b, dtype=dt),
```
Isn't it easier to just do `np.asanyarray(a - b, dtype=unsigned_dt)` - I don't think we're getting any speed-up doing it another way.

seberg replied:
It doesn't work because if `a` or `b` is a negative Python integer, the conversion will raise an error (rather being unsafe and "well defined").

mhvk reviewed (commented):
Looked through the whole PR. I think the main strike against NEP 50 consequences is for comparisons; really, `np.uint*(anything) > -1` should just work. Is the idea right now to decide on that separately?

Looking through the errors: it should be work to do `np.can_cast(np.iinfo(whatever).min, whatever.dtype)` -- maybe those attributes need to be of `whatever.dtype`?

On the percentile failures - I think I'm for option (1) here, ensuring that we keep the weak promotion that is being tested for. It will be really confusing otherwise!

mhvk committed:
Fix percentile and quantile

mhvk commented:
I made a PR to your branch at https://github.com/seberg/numpy/pull/46 - this fixes `percentile` and `quantile`. It all feels a bit fragile, but not horribly so, and it seems best for this PR to not change behaviour that logically is consistent with NEP 50. So, probably the fixes are for the better.

Aside: it also fixes a bug for `longdouble` where the gamma factors were (I think) calculated in regular float rather than double precision.

mhvk commented:
Of course, you are right in https://github.com/seberg/numpy/pull/46/files#r1261584859 that this problem may be more general -- I guess that is an argument for not worrying so much and instead remove those `dtype` expectation tests. It does seem that for scalars at least one really should just not worry too much! What matters is for array-scalar interaction where the output is an array too.

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4119,7 +4119,8 @@ def test_extension_incref_elide_stack(self):
         # check that we don't elide into a temporary which would need casting
         d = np.ones(200000, dtype=np.int64)
-        assert_equal(((d + d) + 2**222).dtype, np.dtype('O'))
+        r = ((d + d) + np.array(2**222, dtype='O'))
```
`2**222` overflows, so would raise.  I would think its well enough tested that it produces `i8`, the test here covers the behavior when overflow happens.

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5006,7 +5007,19 @@ def test_basic(self):
+    @pytest.mark.parametrize("inplace", [False, True])
+    def test_int_range_error(self, inplace):
+        # E.g. clipping uint with negative integers fails to promote
+        # (changed with NEP 50 and may be adaptable)
```
There are possible solution, but right now they need a fair bit of complexity (maybe some adaptation too, but not much).  I think we need a ufunc implementation for `(any_integer, pyinteger)` and `(pyinteger, anyinteger)`.
And either we allow access to the scalar values (if an abstract scalar) in one of the steps (promotion or loop selection), or (which is possible now) we put it into the inner-loop itself.

I.e. you can effectively make it a loop that uses the Python object for the scalar (and then deal with the rest in the loop itself).

That ain't pretty, although fortunately it would only be necessary for comparisons.

(maybe there are other ways to tweak it to be a bit more reasonable, not sure)

seberg commented on numpy/core/tests/test_umath.py:
```diff
@@ -4023,18 +4024,28 @@ def test_float(self):
+        # but promotion rules mean that we try to use a normal integer
+        # in the following case:
+        with pytest.raises(OverflowError):
+            np.equal(2**200, 2**200)
```
Hmmmmmm, some thoughts:
* For comparisons, yes, I think we can just allow this by saying that `PyInt + PyInt -> bool` (i.e. a comparison loop) can just use `object + object -> bool`.  For float or complex it doesn't really matter anyway (well we currently allow complex `<`, but whatever).
* Beyond that, I think things may get tricky:
  - Some functions behave differently for objects or may not have an object loop at all, so I am not sure you can do this choice universally.
  - I might start wondering if we should do this more generally, which then diverges a bit into the whole problem of returning scalars vs. arrays.

Let's focus on the first part (comparisons).  I wonder if it would be very useful if we don't also do the general `int64(2) < 2**200` implementation/hack, but I suppose it may remove some pressure at least from test suits that use NumPy asserts all over.

mhvk replied:
I think it makes sense to remove pain points in order of simplicity!  Certainly, deferring `PyInt, PyInt` to `O,O` for comparisons seems logical. Let's move out further discussion of comparisons to the main thread.

mhvk commented:
Looking again at NEP 50, the main rationale is that values should not determine the result type (correctly!). That also means that for comparisons there is a bit of leeway, since the result type is always bool. And I do think we need to avoid that `np.uint8(1) > -1` raises while `np.uint8(1) > 1.0` works fine... 

One option is to make the behaviour of the comparison operator different from the ufunc (as it is already), and for the operator try to ensure that an error is raised really only when necessary - i.e., let the operator check what is the problem if there is an `OverflowError`. This would make it more like the python comparison operators.

seberg commented:
Right, comparisons are special and it is indeed very reasonable to allow `uint8 > -1` (just not very straight forward unfortunatley); just like I added special `uint64 < int64` comparisons.

Putting a special case into the operator is a good idea!  The worse hack but it should be a straight forward one, at least.

seberg committed:
TST: xfail test that checks `can_cast()` beahvior with pyints

seberg committed:
MAINT: Also adopt `q` handling from the main quantile/percnetile to na version

seberg committed:
MAINT: Address review comments by Marten

seberg committed:
TST: Fix one more tests

Was hidden locally due to longdouble having the same size as double

seberg committed:
TYP: Fix typing passing test which cannot pass anymore

Note that we don't seem to type signed/unsigned mixing for the
cases where they promote to floats correctly, so there cannot be
a new test case.
(This is unrelated to NEP 50, as it already applied to array or scalar
only cases.)

seberg committed:
DOC: Fix docs that cause an integer overflow

seberg committed:
MAINT: Avoid overflow in comparison with int64 on shorter platforms

Maybe not the best way, but its functional and easy to grok and I
doubt it matters speed wise.

seberg committed:
DOC: Fix docstring pointing out quirks that got less bad with NEP 50

seberg commented:
I would like some input, I opened a "tracking issue" for this, but how to approach/continue here?

I think we need to push forward soon and cut corners, because there are too many moving parts for me to keep track off, and I could probably to reduce some of them if we disable `legacy` promotion.  But that requires putting this in.

Right now, this works, the two main problems are (as mentioned many times):
1. `can_cast(123, "int8")` doesn't yet work.  But I am not sure it is used *that* much.
2. Comparison ufuncs with `unsigned > -1` and also `isfinite(integer)` failing.  I am relatively confident we can find a solution for that, but of course merging this might mean downstream has to fix or skip a few tests for moderately good reasons (since we are likely to make things work again).

Of course there is more, many functions in principle should use weak-promotion rules but are not doing so yet due to early `asarray()` calls.

The only test failure (that isn't explicitly `xfailed`) is the `can_cast` documention.

mhvk commented:
@seberg - Ideally, we have a PR that fixes the comparisons before merging this (I don't worry about `can_cast`). Numpy 2.0 changes are quite disruptive already and I think we should try hard to avoid merging something that has known problems that we intend to fix - it will just mean people fixing tests/rewriting code -- most won't know the problems will be fixed and in any case want to have their tests against numpy-dev to pass so that those are useful to find other regressions.

seberg committed:
TST: Undo test changes related to comparising to out-of-bound ints

seberg committed:
MAINT: Undo now unnecessary random number cast to int64

seberg committed:
API: Disable `can_cast` with Python scalars (for now)

seberg committed:
TST: Adjust test to avoid overflow (and bad results on BSD)

Readding invalid warning ignoring, even though it should not matter
(it does seem to for macos arm64 tests, maybe some M2/clang version issue...).

seberg committed:
BENCH: Explicitly skip reversed (arg)partition case because they fail

This doesn't seem like an ideal solution, but is "minimal"?

mhvk commented:
@seberg - the progress sounds good, but at least for me this PR is now far too large to meaningfully review. Would it at all be possible to split it up in pieces? E.g., at least in principle, it would seem that fixing the comparisons does not require enabling NEP 50 behaviour by default. Same for, say, the `quantile` and `percentile` changes.

In order to allow more incremental progress, is there already a CI run that has `NPY_PROMOTION_STATE=weak`? That would probably mean marking quite a few tests with `xfail(weak_promotion)` for now (or make the test dependent on the state if that is relevant), but perhaps helps give a way forward, if only by making clear where the problems are.

seberg commented:
@mhvk I first planned to just put the new stuff into its own PR, but thought I would push it here because I wanted to test the full thing.  I squashed the change, but only to split it out easier into gh-24915.
I would prefer not to try to add a full test run, TBH.  I don't think it helps much with the difficult bigger things, which we can review well enough without trying to keep track of xfails.

mhvk commented on numpy/_core/tests/test_dtype.py:
```diff
@@ -1465,13 +1465,12 @@ def test_complex_pyscalar_promote_rational(self):
-        # or object depending on the input value.  So test those paths!
-        expected_dtype = np.result_type(np.array(val).dtype, np.array(0).dtype)
+        # If we only pass scalars (mainly python ones!), NEP 50 means
+        # that we get either the default integer
```
delete "either"

mhvk commented on numpy/_core/tests/test_multiarray.py:
```diff
@@ -5012,7 +5013,19 @@ def test_basic(self):
+        # Similar to last check in `test_basic`
+        x = (np.random.random(1000) * 255).astype("uint8")
+        with pytest.raises(OverflowError):
+            x.clip(-1, 10, out=x if inplace else None)
```
Hmm, should we change `clip` as well to use `resolve_descriptors_raw`?

seberg replied:
You will have to do a 3 way (you have to check two arguments) dispatching stuff which will be seems like a lot of trouble for a niche feature.  So it seems like a lot of trouble for something very niche?  It isn't even clear to me that the test is testing anything we expect users to do.
(I may be wrong, but I expect more floating point uses and this type of clip call is only relevant when you don't know the input integer dtype and also do not want to cast anyway.)

mhvk replied:
`clip` goes through python code in `_methods.py` that already special-cases `min` and `max` values: https://github.com/numpy/numpy/blob/d885b0b546cbe2ab8401622c8d4db984ec5ea31b/numpy/_core/_methods.py#L92-L101

I made a PR with a possible fix -- https://github.com/seberg/numpy/pull/48 -- though at some level that just brings up the next question: if `clip`, why not also `np.maximum` and `np.minimum` (one-sided only, i.e., `np.minimum(int_array, py_int)` should error if `py_int` is less than the minimum possible given the dtype, but just pass through `int_array` if it is larger than the maximum possible).

In that respect, maybe editing the python function is the right call, since that already special-cases `None` as well.

seberg replied:
Thanks, I was tempted to make it a follow-up if there is an actual complaint.  Do you really think it is important?

Probably only a theoretical caveat, but all of these function (for clip, I think it is a bad choice), return the result type of mixing everything.  That is slightly surprising to get something that can't hold all values.  Although, I have to admit, it isn't really problematic in practice.

mhvk replied:
After I wrote my change, I must admit I became unsure, especially after having to adjust the tests for numpy ints to not do in-place -- as you say, this makes very little sense! I'm definitely fine with just raising an issue that the `clip` dtype resolution needs more work and keeping things as you have them here.

mhvk commented on numpy/_core/tests/test_numeric.py:
```diff
@@ -1426,6 +1422,8 @@ def test_can_cast_structured_to_simple(self):
                                 casting='unsafe'))
 
+    @pytest.mark.xfail(np._get_promotion_state() != "legacy",
+            reason="NEP 50: no int/float/complex (yet)")
```
no *python* int/float/complex, right?

mhvk commented on numpy/_core/tests/test_umath.py:
```diff
@@ -422,13 +422,14 @@ def test_unsigned_signed_direct_comparison(
         assert py_comp(arr, -1) == expected
         assert np_comp(arr, -1) == expected
+
+
```
One empty line...

mhvk commented on numpy/_core/tests/test_umath.py:
```diff
@@ -422,13 +422,14 @@ def test_unsigned_signed_direct_comparison(
         assert py_comp(scalar, -1) == expected
         assert np_comp(scalar, -1) == expected
 
-
```
Two empty lines between classes is good.

seberg replied:
strange, probably some conflict resolution

mhvk commented on numpy/lib/_function_base_impl.py:
```diff
@@ -4461,7 +4460,12 @@ def quantile(a,
-    q = np.asanyarray(q)
+    # Use dtype of array if possible (e.g., if q is a python int or float).
+    if isinstance(q, (int, float)) and a.dtype.kind == "f":
+        q = np.asanyarray(q, dtype=np.result_type(a, q))
```
I'm confused about this: can it not just be `dtype=a.dtype`?

mhvk commented on numpy/lib/_nanfunctions_impl.py:
```diff
@@ -1538,7 +1537,12 @@ def nanquantile(
-    q = np.asanyarray(q)
+    # Use dtype of array if possible (e.g., if q is a python int or float).
+    if isinstance(q, (int, float)) and a.dtype.kind == "f":
+        q = np.asanyarray(q, dtype=np.result_type(a, q))
```
As above, just `dtype=a.dtype`? I know I looked at this before, so probably there was a reason for this... If so, let's add it to the comment...

seberg replied:
Nah, I think it was added for good sport probably.  It might do byte-swaps, etc. but that doesn't matter much.

There may have been a reason: slightly saner user dtype support.  But it seems so niche (and unlikely to affect anything), that I am happy to ignore it for now.

mhvk commented on numpy/typing/tests/data/pass/bitwise_ops.py:
```diff
@@ -63,11 +63,11 @@
-u8 | AR
-u8 ^ AR
-u8 & AR
+i << AR
```
Bit of an illogical place for this. Move up, just after the `i8` set?

mhvk reviewed (commented):
I think with the comparisons done, this is almost all OK. The one ufunc that still is a bit nasty is `clip` - I think it makes sense to use `resolve_descriptors_raw` for it too.

Otherwise, all small comments.

seberg commented on numpy/_core/tests/test_ufunc.py:
```diff
@@ -547,6 +547,9 @@ def test_partial_signature_mismatch_with_cache(self):
             np.add(np.float16(1), np.uint64(2), sig=("e", "d", None))
 
+    @pytest.mark.xfail(np._get_promotion_state() != "legacy",
+            reason="NEP 50 impl breaks casting checks when `dtype=` is used "
```
This is a tricky issue to solve in a generic way, because what we need is a way to define cast safety for scalars generically (additionally it is a bit).

It is on the tracking issue, in most cases I think it is very niche (`dtype=` use is very rare, and this only really matters for `dtype=int` with Python floats input).

You can put a clunky thing in place to make it work for Python ints explicitly, but not sure this isn't so niche that it is a priority?

mhvk replied:
I agree it is super niche; I think the main use case might be where people rely on "random" input to be cast to something as part of an operation (say, `np.substract(a, b, dtype='f8')` when `a` and `b` are passed-in arguments.

Anyway, my comment here was mostly a question of whether if we don't fix it now (which is fine), it is worth raising a separate issue about it so we will fix it eventually

seberg commented on numpy/_core/src/multiarray/multiarraymodule.c:
```diff
@@ -3274,157 +3274,161 @@ PyArray_Where(PyObject *condition, PyObject *x, PyObject *y)
+    ret = (PyObject*)NpyIter_GetOperandArray(iter)[0];
 
-        int has_ref = PyDataType_REFCHK(common_dt);
+    npy_intp itemsize = common_dt->elsize;
```
Sloppy, but can't go wrong I guess.  But changed.

seberg committed:
MAINT: Address review comments

mhvk reviewed (commented):
@seberg - This looked ready to go in (just one small comment, which you can ignore).

To see how things were,  I thought I'd testit with astropy. It seems mostly fine -- there are just a few cases where we need to adjust to ensure casting works fine, mostly for float32. Quite a few of those were corner cases already so no problem. And some work-arounds could be removed.

But I also ran into what I think is a similar problem to what we had with integers for comparisons: `large_py_float > float32` raises an `OverflowError` when it should just be true. I think we should deal with this... (note that I could adjust the piece of code that fails, but it seems wrong in the same way the integers were wrong).

seberg commented:
Hmmmm, interesting and also just some notes for now because I have to sort my thoughts to make up my mind:
* You are seeing  warning turned to error `np.errstate(over="ignore")` will hide the warning. (which means end-users won't notice this much).
* There are some interesting considerations:
  1. By casting float64 to float32, we lose precision, which changes `==` (and thus also `<`).
  2. By casting an out-of-bound float64 to float32 we get an infinity:  This is important because `inf == inf` so we change the bounds of that.

The last point means that the warning actually warns about relevant behavior (although typical user is unlikely to understand of course).

Fixing 2. to get rid of the warning seems like a dubious choice to me unless we also fix 1. which means we introduce explicit `float32` to `float64` comparisons (and `float16`, etc.).
OTOH, I am not sure that fixing 1. is even desired because it means that `np.float32(3.1) == 3.1` fails in some cases.

For integers it was a clear cut thing were it just makes sense to allow as user API and the difficulty is a technical.  But here it is more fuzzy than that!

What does everyone thing?

seberg commented:
@ngoldbaum can you have a (last?) look over.  I would ping everyone else too, but not sure who wants to look at it.

ngoldbaum commented:
Sure, I'll take a look on Monday.

mhvk commented:
The actual comparison that failed (warned) is
```
header[kw] > np.finfo(np.float32).max
```
where `header[kw]` is a float. Getting a warning here is surprising at first! But, as you say logical: while in this case the code gives the "right" answer by casting the python float to `float32` and let it become `inf`, that one gets `True` for `== np.float32(np.inf)` might not be immediately expected. On the other hand, as you note, we want for sure that `3.1 == np.float32(3.1)`, so the cast to `float32` is logical in general. The question then is whether it would make sense to treat the `float64` values that are unrepresentable just because they are outside of the `float32` range differently from those that are unrepresentable because of the precision. 

I think I am now convinced that in the end that would require guessing what the user wants, which makes this unsolvable (and hard to explain). 

The other solution to my specific example would be to make all the attributes of `finfo` return regular `float`. But I think I can see cases where it would be really handy to have the exact representation, so better just keep as is too. After all, in astropy, I can just cast the right side to `float` and solve the issue.

Anyway, for this PR we clearly do not need to solve this. What seems most important is to document this example in perhaps a general casting page...

seberg committed:
MAINT: Clean up error path (DescrFromType cannot fail for simple builtins)

But move the error check to be higher up and thus more directly
relevant.  If `PyArray_DescrFromType` could have failed, the code
would have been incorrect anyway.

mhvk commented on numpy/_core/src/multiarray/multiarraymodule.c:
```diff
@@ -3297,15 +3297,14 @@ PyArray_Where(PyObject *condition, PyObject *x, PyObject *y)
     NpyIter * iter;
     NPY_BEGIN_THREADS_DEF;
 
-    if (common_dt == NULL || op_dt[1] == NULL) {
```
I think you still need to check `op_dt[1] != NULL` right after the definition of `op_dt` - though really that should not fail...

mhvk replied:
OK, makes sense! If it fails, there are certainly worse problems afoot...

mhvk reviewed (approved):
It would be great for @ngoldbaum to have a last look, but in the meantime let me explicitly approve, since I think this is ready to go in...

ngoldbaum commented on numpy/_core/multiarray.py:
```diff
@@ -570,52 +572,6 @@ def can_cast(from_, to, casting=None):
-    False
-
-    >>> np.can_cast('<i8', '>u4', 'unsafe')
-    True
```
Should we leave some examples behind that demonstrate the new semantics for numpy scalars?

seberg replied:
It just raises a `TypeError` right now, so not sure it is very useful?

ngoldbaum replied:
Ah, that wasn't clear, never mind.

ngoldbaum commented on numpy/_core/src/multiarray/multiarraymodule.c:
```diff
@@ -3252,8 +3252,9 @@ array_set_datetimeparse_function(PyObject *NPY_UNUSED(self),
 PyArray_Where(PyObject *condition, PyObject *x, PyObject *y)
 {
-    PyArrayObject *arr, *ax, *ay;
+    PyArrayObject *arr, *ax, *ay = NULL;
```
Might as well initialize all of these to `NULL`, leaving it like this is a little confusing

seberg replied:
Added, although I can't say I prefer it when it isn't needed for cleanup (the "uninitialized" warnings work pretty well anyway).

ngoldbaum commented on numpy/_core/src/multiarray/multiarraymodule.c:
```diff
@@ -3274,161 +3275,163 @@ PyArray_Where(PyObject *condition, PyObject *x, PyObject *y)
+    if (common_dt == NULL) {
+        goto fail;
+    }
+    PyArray_Descr * op_dt[4] = {common_dt, PyArray_DescrFromType(NPY_BOOL),
```
Couldn't hurt to add a `\* creating a built in dtype like this cannot fail *\` above this line.

seberg replied:
OK, we really should probably just have a name for it...

ngoldbaum commented on numpy/_core/src/multiarray/multiarraymodule.c:
```diff
@@ -3274,161 +3275,163 @@ PyArray_Where(PyObject *condition, PyObject *x, PyObject *y)
     ay = (PyArrayObject*)PyArray_FROM_O(y);
-    if (ax == NULL || ay == NULL) {
+    if (ay == NULL) {
         goto fail;
```
Sorry for writing this with awkward control flow in the first place... :upside_down_face:

ngoldbaum commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -935,6 +935,19 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
+             *       isnan).  This is not ideal, but pragmatic...
+             *       We should eventually have special loops for isnan and once
+             *       we do, we may just deprecate all remaining ones (e.g.
+             *       `negative(2**100)` not working as it is an object.)
```
It would be nice if this comment mentioned it was a TODO from the NEP 50 work, for context

seberg replied:
I have it in the tracking issue...

ngoldbaum replied:
I'm thinking of a reader who doesn't have context about NEP 50 and wouldn't be aware of a tracking issue.

seberg replied:
Ah, sorry, somehow was thinking about creating an issue.

ngoldbaum commented on numpy/_core/tests/test_numeric.py:
```diff
@@ -1039,20 +1039,26 @@ def check_promotion_cases(self, promote_func):
-        assert_equal(promote_func(i64, np.array([u32])), np.dtype(np.uint32))
+        assert_equal(promote_func(np.array([i8]), i64), np.dtype(np.int64))
+        # unsigned and signed unfortunately tend to promote to float64:
+        assert_equal(promote_func(u64, np.array([i32])), np.dtype(np.float64))
```
Sorry for coming into this super late - this seems like a surprising corner case. I guess this was discussed before though?

seberg replied:
The PR doesn't change the behavior at all, it just before this was the same as `python int == int32`, the mixing was always strange like that.

ngoldbaum reviewed (commented):
I saw a few places that could be clarified a bit and I have a couple questions.

seberg commented on doc/source/user/basics.creation.rst:
```diff
@@ -43,13 +43,18 @@ When you use :func:`numpy.array` to define a new array, you should
 
-  >>> a = np.array([127, 128, 129], dtype=np.int8)
-  >>> a
+  >>> np.array([127, np.int64(128), np.int64(129)], dtype=np.int8)
```
OK, removed it.  I agree, I don't like it much, I just didn't really want to remove ahe note about overflows being possible completely.

seberg committed:
MAINT: Address Nathans review comments

ngoldbaum committed:
MAINT: appease linter
